vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +556,603 @@ void free_route_tree(t_rt_node * rt_node) {
+	 *	It is the **caller's responsibility** to actually 					*
+	 *	call pathfinder_update_cost and free_route_subtree 					*
+	 *																		*
+	 *	The removal is only done at branch points for its branches 			*
+	 *	returns whether the entire tree was pruned or not 					*
+	 *	assumes arriving rt_root does not have its R_upstream calculated 	*
+	 *	Tdel for the entire tree after this function is still unset, 		*
+	 *	call load_route_tree_Tdel(rt_root, 0) at SOURCE 					*/
+
+	assert(rt_root != nullptr);
+
+	auto inode = rt_root->inode;
+
+	// illegal tree, propagate information back up (actual removal done at an upstream branch point)
+	if (rr_node[inode].get_occ() > rr_node[inode].get_capacity()) return true;
+	// legal routing, allowed to do calculations now
+
+	// can calculate R_upstream from just upstream information without considering children
+	auto parent_switch = rt_root->parent_switch;
+	if (g_rr_switch_inf[parent_switch].buffered)
+		R_upstream = g_rr_switch_inf[parent_switch].R + rr_node[inode].R;
+	else
+		R_upstream += g_rr_switch_inf[parent_switch].R + rr_node[inode].R;
+	rt_root->R_upstream = R_upstream;
+
+	auto edge = rt_root->u.child_list;
+	// no edge means it must be a sink; legally routed to sink!
+	if (!edge) {
+		reached_rt_sinks.push_back(rt_root);
+		rt_root->C_downstream = 0;	// sink has no node downstream and always has 0 own capacitance
+		return false;
+	}
+	
+	// prune children and get their C_downstream
+	float C_downstream_children {0};
+
+	t_linked_rt_edge* prev_edge {nullptr};
+	do {
+		bool pruned {prune_illegal_branches_from_route_tree(edge->child, pres_fac, reached_rt_sinks, remaining_targets, R_upstream)};
+		// need to remove edge and possibly change first edge if pruned
+		if (pruned) {
+			// optimization for long paths (no sibling edges), propagate illegality up to branch point
+			if (!rt_root->u.child_list->next) {
+				return true;
+			}
+			// else this is a branch point and the removal should be done here
+			find_sinks_of_route_tree(edge->child, remaining_targets);
+			edge = free_route_subtree(rt_root, prev_edge);	// returns next edge
+			// previous edge does not change when the current edge gets cut
+		}
+		else {
+			// child still exists and has calculated C_downstream
+			if (!g_rr_switch_inf[edge->iswitch].buffered)
+				C_downstream_children += edge->child->C_downstream;
+			prev_edge = edge;
+			edge = edge->next; // advance normally as branch still exists
+		}
+	} while (edge);
+	// entirely pruned, this should never be reached due to propagating illegality up to branch point for paths
+	assert(rt_root->u.child_list != nullptr);
+
+	// the sum of its children and its own capacitance
+	rt_root->C_downstream = C_downstream_children + rr_node[inode].C;
+	return false;
+}
+bool prune_route_tree(t_rt_node* rt_root, float pres_fac, vector<t_rt_node*>& reached_sinks, vector<int>& remaining_targets) {
+
+	/* Prune a skeleton route tree of illegal branches (in terms of congestion)						  *
```

Comment:
probably define illegal branches explicitly rather than "in terms of congestion". 
"in terms of congestion" could have the meaning that branches are only illegal if it is *highly* congested.